protocol/test-serialization/src/fuzz.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(test)]
+mod tests {
+    use crate::fuzz_test;
+    use bin_prot::Deserializer;
+    use mina_rs_base::types::*;
+    use rand::prelude::*;
+    use serde::Deserialize;
+    use wasm_bindgen_test::*;
+
+    #[test]
+    #[wasm_bindgen_test]
+    fn test_corrupted_deserialization() {
+        fuzz_test!(
+            ExternalTransition
+
+            ProtocolState
+            ProtocolStateBody
+            BlockchainState
+            ConsensusState
+            ProtocolConstants
+
+            ProtocolStateProof
+            ProofStatement
+            PrevEvals
+            Proof
+            ProofMessages
+            ProofOpenings
+
+            StagedLedgerDiff
+            StagedLedgerDiffTuple
+            StagedLedgerPreDiffTwo
+
+            DeltaTransitionChainProof
+
+            Option<ProtocolVersion>
+            ()
+        );
+    }
+
+    #[macro_export]
+    macro_rules! fuzz_test {
+        ($($ty: ty) *) => {
+            $(
+                let mut rng = rand::thread_rng();
+                for _i in 0..5 {
+                    let mut bytes = vec![0; rng.gen_range((1024 * 1024)..(50 * 1024 * 1024))];
```

Comment:
This `gen_range` is producing a large amount of bytes, bigger than any type.

This is a good test, you should probably also have some tests where the size of the random bytes is capped at the size of the corresponding object you are deserialising?  This way we know the test is not just passing because of a size check.

Reply:
The binary is deserialized as a stream and in the current design, it does not require the stream to reach its end to allow non-exhaustive struct definition(although the omitted field has to be the tailing ones), I've added another `ensure_non_exhaustive_deserialization` test to demonstrate that.

Reply:
Thanks 👍🏼